### ISSUE TITLE
Linker conflicts and usage clarification [clang]

### ISSUE THREAD

**Issue by henningkayser** [ros2/console_bridge_vendor#19]
We're currently running into linker conflicts when compiling MoveIt 2 for Foxy with clang.

* Travis: https://travis-ci.com/github/ros-planning/moveit2/jobs/372952760#L578
* MoveIt Issue: https://github.com/ros-planning/moveit2/issues/242
* Original comment: https://github.com/ros-planning/geometric_shapes/pull/153#discussion_r465706379

```
/usr/bin/ld: warning: libconsole_bridge.so.0.4, needed by /opt/ros/foxy/lib/x86_64-linux-gnu/liburdfdom_sensor.so.1.0, may conflict with libconsole_bridge.so.1.0
```

MoveIt is depending on `console_bridge` through [geometric_shapes](https://github.com/ros-planning/geometric_shapes/blob/ros2/package.xml#L30) and [srdfdom](https://github.com/ros-planning/srdfdom) which recursively depends on [urdfdom](https://github.com/ros2/urdfdom). The problem is that the `urdfdom` debian (and many other ROS packages) depends on `libconsole_bridge.so.0.4` while `console_bridge_vendor` already provides `libconsole_bridge.so.1.0` resulting in the linker conflict shown above. My understanding was that this is exactly the issue that the vendor package should solve. However, if using the vendor package or not doesn't seem to make any difference when compiling `srdfdom` and MoveIt 2 (see https://github.com/ros-planning/srdfdom/pull/61/files for `srdfdom` tested with and without vendor package). The linker conflict doesn't occur when compiling the conflicting packages (here `urdfdom`) from source. To me it seems like the only option would be to have `urdfdom` released using the latest `console_bridge` version and from then on keeping all downstream dependencies synced.

Is this by any chance just a bug in the vendor package or a misconfiguration in one of the related packages? Any help would be greatly appreciated.

@clalancette tagging you as you've been involved in this package and in the original discussion.

**Comment by henningkayser** [comment]
@Blast545 @ahcorde @jacobperron any insights?

**Comment by jacobperron** [comment]
Looking at the CMakeLists.txt of the vendor package, it requires console_bridge version `1.0.1`:

https://github.com/ros2/console_bridge_vendor/blob/6cffe405a707a4cbf6ff7a1d13ac22552b3e22ae/CMakeLists.txt#L73

Before that, it was requiring version `0.5.4`:

https://github.com/ros2/console_bridge_vendor/blame/2545ecd578fa80e7bc44c88b714239676e667b91/CMakeLists.txt#L73

I'm guessing it's not a bug that we are requiring a newer version of console_bridge, though I don't know the reasons for relying on a newer version.

I think depending on `console_bridge_vendor` is the correct approach. It is our way to use newer versions of the package since getting new releases upstream in Debian can take longer than some of our iterations. To me, it looks like the bug is in srdfdom, which should be fixed by the referenced PR (https://github.com/ros-planning/srdfdom/pull/61).

**Comment by henningkayser** [comment]
Thanks for clarification, that's what we'll do then. Do you see any way to avoid the linker conflict above which seems to occur by the `urdfdom` debian still using `libconsole_bridge.so.0.4`? Or do we actually have to compile `urdfdom` from source until there is a new release?

**Comment by jacobperron** [comment]
@henningkayser Are you able to depend on the ROS distro version of `urdfdom` instead? E.g. `ros-foxy-urdfdom`

AFAIK, that should be using the version of console_bridge installed by the vendor package.

**Comment by jacobperron** [comment]
To clarify my comment above, I recommend installing the version of `urdfdom` shipped with the ROS distro. As long as you declare a dependency in your `package.xml` to `urdfdom`, then `rosdep` should handle the rest.

**Comment by henningkayser** [comment]
@jacobperron That's exactly what we're doing. The described issue only happens when using the ROS distro debian `ros-foxy-urdfdom`. The command `apt-cache depends ros-foxy-urdfdom` lists `libconsole-bridge0.4`  and also `ros-foxy-console-bridge-vendor`, so it actually seems to depend on two different console_bridge versions at the same time (`ros-foxy-console-bridge-vendor` has been updated in the meantime). When installing `urdfdom` from source, the conflict doesn't happen because it just picks the latest console_bridge version from the vendor package.

**Comment by jacobperron** [comment]
@henningkayser Interesting. Sorry, I misunderstood the first time.

Can you provide some simple steps to reproduce the issue? I've tried building the moveit2 packages (minus the urdfdom and urdfdom_headers packages) against a Foxy binary installation, but haven't reproduced.

**Comment by clalancette** [comment]
@henningkayser Friendly ping here, we're having trouble reproducing.  Can you provide more information?  Thanks.

**Comment by henningkayser** [comment]
@clalancette @jacobperron thanks for looking into this. For reproducing this you should just have to follow the install instructions and remove `urdfdom` and `urdfdom_headers` from your workspace (use the debians instead) and compile with clang. For me (and for [Travis](https://travis-ci.com/github/ros-planning/moveit2/jobs/372952760#L602)) this reliably results in the described behavior. With `gcc` this issue only seems to occur occasionally.

**Comment by clalancette** [comment]
I've done some looking here, and I've been able to reproduce the problem that `urdfdom` links to the wrong version of `console_bridge` *sometimes*.  It's clear that it did it on the buildfarm: http://build.ros2.org/view/Fbin_uF64/job/Fbin_uF64__urdfdom__ubuntu_focal_amd64__binary/30/consoleFull .  I've also determined that for the ROS 2 core, the packages that are currently affected by this are `rviz_default_plugins`, `urdf`, `urdfdom`, `robot_state_publisher`, `kdl_parser`, and `qt_gui_cpp`.

If I force `urdfdom`'s CMakeLists.txt to look for console_bridge 1.0, it does the right thing:

```
find_package(console_bridge 1 REQUIRED)
```

What I have not yet been able to figure out is if we don't specify the version, it sometimes picks up the system one, and sometimes picks up the one in `/opt/ros/foxy`.  They are clearly both available, I just don't know how CMake chooses the order in which to search.

**Comment by clalancette** [comment]
And just to follow-up; the root problem is that the buildfarm doesn't properly set CMAKE_PREFIX_PATH so it searches for vendor packages in the order we want.  The most correct solution (and the one we should do going forward) is to patch bloom to add `CMAKE_PREFIX_PATH` as appropriate to the debian builds, which matches what `colcon` does when building locally.  We're also doing something similar with the RPM builds currently: https://github.com/ros-infrastructure/bloom/blob/master/bloom/generators/rpm/templates/ament_cmake/template.spec.em#L47 .  Once we have that, we can re-bloom urdfdom (and any other packages that have this particular issue).

However, a change like this is not without risk.  We still have to determine how to mitigate that risk.

**Comment by henningkayser** [comment]
@clalancette thanks for clarifying the issue. It's really helpful to know that it's actually setting the `CMAKE_PREFIX_PATH` can fix this. What kind of risks are you expecting from this? As long as affected packages don't specify an old version of `console_bridge` I thought this should work just fine.

**Comment by clalancette** [comment]
> @clalancette thanks for clarifying the issue. It's really helpful to know that it's actually setting the `CMAKE_PREFIX_PATH` can fix this. What kind of risks are you expecting from this? As long as affected packages don't specify an old version of `console_bridge` I thought this should work just fine.

Essentially, any change to bloom is risky in that it changes how *all* packages (in all ROS 1 and ROS 2 distributions) are built.  It seems like it should be safe (and more closely mimics what colcon does), but when you spread it across ~6 Linux distributions times 6 ROS distributions times thousands of packages, the risk of breaking one of those combinations goes way up.

**Comment by dirk-thomas** [comment]
TL;DR This is a bug in the templates used by `bloom` to generate the Debian rules file.

First some context to understand why / how this is happening. The `CMAKE_INSTALL_PREFIX` is always added to the search path of e.g. `find_package()` (see [CMake docs](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html)):

> The installation prefix is also added to `CMAKE_SYSTEM_PREFIX_PATH` so that `find_package()`, `find_program()`, `find_library()`, `find_path()`, and `find_file()` will search the prefix for other software.

But the install prefix if **not** inserted before global paths like `/usr`. So the `CMAKE_SYSTEM_PREFIX_PATH` will look like this: `/usr/local;/usr;/;/usr;${CMAKE_INSTALL_PREFIX};/usr/X11R6;/usr/pkg;/opt`. That means if no CMAKE_PREFIX_PATH is set pointing to `/opt/ros/<distro>` that packages are first looked for in global locations like `/usr` before `CMAKE_INSTALL_PREFIX` (which is also `/opt/ros/<distro>`).

Now to the content of the bloom templates - specifically the ones for the Debian rules file (rather than rpm referenced above):
* For `catkin` packages `CMAKE_PREFIX_PATH` is set explicitly: https://github.com/ros-infrastructure/bloom/blob/d2265fd73f7c93437dc9fb704e893f6ef478b8ae/bloom/generators/debian/templates/catkin/rules.em#L29-L32 (in case the sourced setup file doesn't exist since that would set it too)
* For `ament_cmake` packages `CMAKE_PREFIX_PATH` is **not** set explicitly: https://github.com/ros-infrastructure/bloom/blob/d2265fd73f7c93437dc9fb704e893f6ef478b8ae/bloom/generators/debian/templates/ament_cmake/rules.em#L29-L31 (the the sourced setup file never sets the variable either)
* For `cmake` packages `CMAKE_PREFIX_PATH` is also not set explicitly: https://github.com/ros-infrastructure/bloom/blob/d2265fd73f7c93437dc9fb704e893f6ef478b8ae/bloom/generators/debian/templates/cmake/rules.em#L29

Therefore I think `CMAKE_PREFIX_PATH` must be set in both `ament_cmake` as well as `cmake` packages to ensure packages are first searched for in the install prefix and only secondary in system paths.

**Comment by henningkayser** [comment]
@dirk-thomas thanks for getting this bug fixed. I assume we now just need a new release of the affected packages to apply https://github.com/ros-infrastructure/bloom/pull/606 in order to resolve this issue.

**Comment by clalancette** [comment]
> @dirk-thomas thanks for getting this bug fixed. I assume we now just need a new release of the affected packages to apply [ros-infrastructure/bloom#606](https://github.com/ros-infrastructure/bloom/pull/606) in order to resolve this issue.

That's correct.  I did this yesterday for Rolling as a test bed, and it seems to have fixed the issue there.  I'll do similar releases for Foxy today.

**Comment by clalancette** [comment]
All right, the latest Foxy packages (in testing) all seem to be linking against `libconsole_bridge.so.1.0`.  @henningkayser please test it out and make sure it works for you.  I'm going to close this out, but feel free to reopen if it still a problem.

**Comment by ros-discourse** [comment]
This issue has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2020-10-15/16849/1
